### PR TITLE
 Add peer authorization check to ingress connection

### PR DESCRIPTION
**Pull request by kfirtoledo**
Add Peer authorization check to ingress connection. 
Small fix to openspeedtest for checking access policy related to source peer.

### CODE REVIEW

File: pkg/controlplane/server/http/authz.go [hunk 1]
```diff
@@ -43,7 +43,13 @@ func (s *Server) PeerAuthorize(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	resp, err := s.cp.AuthorizeIngress(&controlplane.IngressAuthorizationRequest{Service: req.Service})
+	peer, err := s.cp.AuthorizePeer(r.TLS.PeerCertificates[0].DNSNames)
```
Comment (orozery) [inline pkg/controlplane/server/http/authz.go:46]: The purpose of the `Server` struct is just parsing HTTP and passing it on to the lower layer (which is `controlplane::Instance`).
Thus, my view is that the new `AuthorizePeer` function should be merged to `AuthorizeIngress`.

**Review by zivnevo — APPROVED** [review]
LGTM

**Review by orozery — CHANGES_REQUESTED** [review]

File: pkg/controlplane/authz.go [hunk 2]
```diff
@@ -121,8 +122,20 @@ func (cp *Instance) AuthorizeEgress(req *EgressAuthorizationRequest) (*EgressAut
...
+// AuthorizePeer authorizes a peer for accessing an exported service.
+func (cp *Instance) AuthorizePeer(names []string) (*store.Peer, error) {
+	for _, name := range names {
+		peer := cp.GetPeer(name)
+		if peer != nil {
+			return peer, nil
+		}
+	}
+
+	return nil, fmt.Errorf("peer does not exist")
```
[the remark this answers is not in the document]
Comment (kfirtoledo) [reply pkg/controlplane/authz.go:134]: I think we should limit peer access and not allow anyone in the fabric to access this peer, on the other hand, we can say this is the responsibility of the policy engine.

[the remark this answers is not in the document]
Comment (orozery) [reply pkg/controlplane/authz.go:134]: Only peers which were signed by the root CA can establish TLS connection.
For more fine-tuned access control we have the policy engine.

File: pkg/controlplane/authz.go [hunk 3]
```diff
@@ -121,8 +122,20 @@ func (cp *Instance) AuthorizeEgress(req *EgressAuthorizationRequest) (*EgressAut
 	return resp, nil
 }
 
+// AuthorizePeer authorizes a peer for accessing an exported service.
+func (cp *Instance) AuthorizePeer(names []string) (*store.Peer, error) {
+	for _, name := range names {
```
[the remark this answers is not in the document]
Comment (orozery) [reply pkg/controlplane/authz.go:127]: Our controlplane/dataplane currently verify at startup that the DNS names match the exact format generated by our cl-adm.
Anyhow, I think we talked about requiring the peer CA to be able to sign only DNS names of the form *<peername>.

**Comment by kfirtoledo** [comment]
@orozery PTAL

File: pkg/controlplane/server/http/authz.go [hunk 4]
```diff
@@ -43,7 +43,14 @@ func (s *Server) PeerAuthorize(w http.ResponseWriter, r *http.Request) {
...
 	}
 
-	resp, err := s.cp.AuthorizeIngress(&controlplane.IngressAuthorizationRequest{Service: req.Service})
+	if r.TLS == nil || len(r.TLS.PeerCertificates) == 0 || len(r.TLS.PeerCertificates[0].DNSNames) != 2 ||
+		r.TLS.PeerCertificates[0].DNSNames[0] == "" {
+		http.Error(w, fmt.Errorf("certificate does not contain a valid DNS name for the peer gateway").Error(), http.StatusBadRequest)
+		return
+	}
+
+	peerName := r.TLS.PeerCertificates[0].DNSNames[0]
```
Comment (orozery) [inline pkg/controlplane/server/http/authz.go:52]: This code can be useful for other purposes.
I think I will open-up a followup PR which moves it to `pkg/certs` (with code moved from `pkg/bootstrap/cert.go`).

**Review by orozery — APPROVED** [review]
I

**Pull request merged**